File: parzen.py
from collections import defaultdict

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from tqdm import tqdm

"""
研一下学期的代码，2017年上半年

那时候numpy都用的不熟
"""


def kernel_uniform(u, h, dim):  # 矩形窗核函数，均匀分布
    return 1 / (h ** dim) if u <= 1 / 2 else 0


def kernel_gauss(u, h):  # 高斯核函数，mu和sigma
    return np.exp(-u * u / (2 * h * h)) / (np.sqrt(2 * np.pi) * h)


def norm_max(vec):  # 向量的无穷范数
    return np.max(np.abs(vec))


def norm2(vec: np.ndarray):  # 向量的2范数
    return np.linalg.norm(vec)


def load_letter_recognition(test_size: float):
    # test_size：测试集的比例
    data = pd.read_csv("letter-recognition.data").values
    train_data, test_data = train_test_split(data, test_size=test_size)
    a = defaultdict(lambda: [])  # 把测试数据按照类别分开
    for i in train_data:
        a[i[0]].append(i[1:])
    return a, test_data


def get_ans(test_sample, train_data, h, kernel, norm):
    ans = None  # 目标类别
    ansP = 0  # 目标类别的概率
    test_vec = test_sample[1:]
    for category, samples in train_data.items():  # j中的数据都是同一类别
        s = sum([kernel(norm(sample - test_vec), h) for sample in samples])
        if s > ansP:
            ansP = s
            ans = category
    return ans


def run():
    train_data, test_data = load_letter_recognition(test_size=0.2)
    print('训练集', train_data.keys(), test_data.shape)
    right_cnt = 0  # 正确的个数
    h = 3  # 窗口大小，窗口大小拍脑袋决定不太好
    for i in tqdm(test_data):
        # ans = get_ans(i, train_data, h, kernel_uniform, norm_max)
        ans = get_ans(i, train_data, h, kernel_gauss, norm2)
        if ans == i[0]:
            right_cnt += 1
    print('正确率', right_cnt / len(test_data))


if __name__ == '__main__':
    run()

File: parzen.cpp
#include<stdio.h>
#include<math.h>  
#include<time.h>
const double pi = acos(-1.0);
struct Sample{
	char c;
	int a[16];
}data[20000];
double norm_max(Sample&x, Sample&y){
	double s = 0;
	for (int i = 0; i < 16; i++){
		double now = abs(x.a[i] - y.a[i]);
		if (now>s)s = now;
	}
	return s;
}
double norm2(Sample&x, Sample&y){
	double s = 0;
	for (int i = 0; i < 16; i++){
		double t = x.a[i] - y.a[i];
		s += t*t;
	}
	return s;
}
double kernel_rectangle(double x, double h, int dim){
	return x / h <= 0.5 ? 1 / pow(h, dim) : 0;
}
double kernel_circle(double x, double h){
    //球形空间
	return x <= h ? 1 : 0;
}
double kernel_gauss(double x, double h){
	return exp(-x*x / (2 * h*h)) / sqrt(2 * pi) / h;
}
double kernel_zhishu(double x, double h){
	return exp(-x / h);
}
int main(){
	FILE *infile = fopen("letter-recognition.data", "r");
	if (infile == NULL){
		puts("缺少数据文件");
		return -1;
	}
	for (int i = 0; i < 20000; i++){
		fscanf(infile, "\n%c", &data[i].c);
		for (int j = 0; j < 16; j++){
			fscanf(infile, ",%d", &data[i].a[j]);
		}
	}
	int beg_time = time(0);
	int test_size = 4000, train_size = 2e4 - test_size;
	Sample*train = data, *test = &data[train_size];
	int right_cnt = 0;
	for (int i = 0; i < test_size; i++){
		double p[26] = { 0 };
		for (int j = 0; j < train_size; j++){
			//p[train[j].c - 'A'] += kernel_rectangle(norm2(test[i], train[j]),h,16);
			//p[train[j].c - 'A'] += kernel_circle(norm2(test[i], train[j]), 15);
			p[train[j].c - 'A'] += kernel_gauss(norm2(test[i], train[j]), 3.7);
			//高斯核，h=3.7，正确率96.2%
			//p[train[j].c - 'A'] += kernel_zhishu(norm2(test[i], train[j]), 1.5);
		}
		char ans = 'A';
		double ansP = p[0];
		for (int j = 1; j < 26; j++){
			if (p[j]>ansP){
				ansP = p[j];
				ans = j + 'A';
			}
		}
		if (ans == test[i].c){
			right_cnt++;
		}
	}
	int end_time = time(0);
	printf("耗时%d秒，正确率为%lf\n", end_time - beg_time, right_cnt*1.0 / test_size);
	return 0;
}